File: gaddlemaps/_transform_molecule.py
# -*- coding: utf-8 -*-
#    Gaddlemaps python module.
#    Copyright (C) 2019-2021 José Manuel Otero Mato, Hadrián Montes Campos, Luis Miguel Varela Cabo
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU Affero General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU Affero General Public License for more details.
#
#    You should have received a copy of the GNU Affero General Public License
#    along with this program.  If not, see <https://www.gnu.org/licenses/>.
'''
This module contains functions that change the molecules conformation.
'''


from collections import deque
from typing import Dict, List, Tuple, Deque

import numpy as np


def move_mol_atom(atoms_pos: np.ndarray,
                  bonds_info: Dict[int, List[Tuple[int, float]]],
                  atom_index: int = None,
                  displ: np.ndarray = None,
                  sigma_scale: float = 0.5) -> np.ndarray:
    """
    Moves an atom of a molecule respecting almost all bond distances.

    By default, a random atom is picked from atom_pos and moved randomly in a
    certain direction (see the published article for a better description of
    this step).

    Parameters
    ----------
    atoms_pos : numpy.ndarray
        An array with the positions of the molecule atoms in rows.
    bonds_info : dictionary
        A dict with the information of the bonds. Example:
            bonds_info = {0:[(1, 5.4), (2, 6.4)], 1:[(0, 5.4), ], ...}
        The keys refers to atom index and values are lists with tuples. Each
        tuple contains the bonded atom index and the bond length.
    atom_index : integer (Optional)
        The index of the atom to move (respecting the index of atoms_pos). If
        None is given a random one is taken.
    displ : numpy.ndarray (Optional)
        The displacement vector. If None is given a random displacement is
        calculated in the normal plane to the line jointing most nearest atoms.
    sigma_scale : float
        A factor to scale the sigma of the distribution of the
        displacement module.

    Returns
    -------
    modified_atoms_pos : numpy.ndarray
        An array with the modified positions of the atoms in rows.

    """

    atoms_pos = np.copy(atoms_pos)
    n_atoms = len(atoms_pos)
    if atom_index is None:
        atom_index = np.random.randint(n_atoms)

    if displ is None:
        displ = find_atom_random_displ(atoms_pos, bonds_info, atom_index,
                                       sigma_scale=sigma_scale)
    # Create a queue with the atoms to move to restore the bonds
    wait_queue = deque(range(n_atoms))
    # Remove an atom index if it was already moved
    atoms_pos[atom_index] += displ
    wait_queue.remove(atom_index)

    queue: Deque[Tuple[int, int, int]] = deque()
    for i in bonds_info[atom_index]:
        queue.append((atom_index, i[0], i[1]))  # type: ignore
        wait_queue.remove(i[0])

    while queue:
        ind1, ind2, bond = queue.pop()
        diferencia = atoms_pos[ind1] - atoms_pos[ind2]
        modulo = np.linalg.norm(diferencia)
        unit = diferencia/modulo
        atoms_pos[ind2] = atoms_pos[ind2] + (modulo - bond) * unit
        for bonds in bonds_info[ind2]:
            if bonds[0] in wait_queue:
                queue.append((ind2, bonds[0], bonds[1]))  # type: ignore
                wait_queue.remove(bonds[0])
    return atoms_pos


def find_atom_random_displ(atoms_pos: np.ndarray,
                           bonds_info: Dict[int, List[Tuple[int, float]]],
                           atom_index: int,
                           sigma_scale: float = 0.5) -> np.ndarray:
    """
    Finds a random displacement for the atom with a given index.

    This displacement is chosen in a perpendicular direction according to the
    number of bonded atoms.

    Parameters
    ----------
    atoms_pos : numpy.ndarray
        An array with the positions of the molecule atoms in rows.
    bonds_info : dictionary
        A dict with the information of the bonds. Example:
            bonds_info = {0:[(1, 5.4), (2, 6.4)], 1:[(0, 5.4), ], ...}
        The keys refers to atom index and values are lists with tuples. Each
        tuple contains the bonded atom index and the bond length.
    atom_index : integer (Optional)
        The index of the atom to move (respecting the index of atoms_pos). If
        None is given a random one is taken.
    sigma_scale : float
        A factor to scale the sigma of the distribution of the displacement
        module.

    Returns
    -------
    displ : numpy.ndarray
        The displacement vector to sum to the position of the interest atom.

    """
    n_bonded_ref = len(bonds_info[atom_index])
    # Calc the width of the displacements distribution
    sigma = bonds_info[atom_index][0][1] * sigma_scale

    if n_bonded_ref == 1:
        direction = np.cross(np.random.rand(3),
                             atoms_pos[bonds_info[atom_index][0][0]] -
                             atoms_pos[atom_index])
    elif n_bonded_ref == 2:
        direction = np.cross(np.random.rand(3),
                             atoms_pos[bonds_info[atom_index][0][0]] -
                             atoms_pos[bonds_info[atom_index][1][0]])
    elif n_bonded_ref >= 3:
        direction = np.cross(atoms_pos[bonds_info[atom_index][0][0]] -
                             atoms_pos[bonds_info[atom_index][2][0]],
                             atoms_pos[bonds_info[atom_index][0][0]] -
                             atoms_pos[bonds_info[atom_index][1][0]])
        direction *= np.random.choice([-1, 1])
    direction = direction / np.linalg.norm(direction)
    displ = direction * np.random.normal(0, sigma)
    return displ

File: cython_backend/c++_src/transform_molecule.cpp
//    Gaddlemaps python module.
//    Copyright (C) 2019-2021 José Manuel Otero Mato, Hadrián Montes Campos, Luis Miguel Varela Cabo
//
//    This program is free software: you can redistribute it and/or modify
//    it under the terms of the GNU Affero General Public License as published by
//    the Free Software Foundation, either version 3 of the License, or
//    (at your option) any later version.
//
//    This program is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//    GNU Affero General Public License for more details.
//
//    You should have received a copy of the GNU Affero General Public License
//    along with/ this program.  If not, see <https://www.gnu.org/licenses/>.
#include "transform_molecule.h"
auto random_engine2 = std::default_random_engine();
auto random_direction = std::uniform_int_distribution<int>(0,1.);
auto random_normal = std::normal_distribution<double>(0.,1.);

void move_mol_atom(arma::mat &atom_pos,
                   bonds &bonds_info,double sigma_scale){
    int Natoms = int(atom_pos.n_rows);
    auto int_dist = std::uniform_int_distribution<int>(0, Natoms-1);
    // Choose one random atom to move
    int atom_index = int_dist(random_engine2);
    // Choose a random displacement
    auto disp = find_atom_random_displ(atom_pos, bonds_info,
                                       atom_index, sigma_scale);

    // Start an empty set of the atoms to move
    std::set<int> to_move;
    // Poulate the queue
    for (int atom=0; atom < Natoms; ++atom){
        to_move.insert(atom);
    }
    // Move the first atom
    atom_pos.row(atom_index) += disp;
    // Remove the value from the queue
    to_move.erase(atom_index);

    // Create the waiting list of the atoms to move
    std::queue<std::pair<int, std::pair<int, double>>> waiting;

    auto bonds_list = bonds_info[atom_index];
    for (int i=0; i < bonds_list.size(); ++i){
        waiting.push(std::make_pair(atom_index, bonds_list[i]));
        to_move.erase((bonds_list[atom_index].first));
    }

    while (! waiting.empty()){
        // Get the info of the priority atom to move
        auto bond_move = waiting.front();
        int ind1 = bond_move.first;
        int ind2 = bond_move.second.first;
        double bond = bond_move.second.second;
        waiting.pop();

        // Move the atom to fix the distance
        arma::rowvec diferencia = atom_pos.row(ind1) - atom_pos.row(ind2);
        double modulo = arma::norm(diferencia);
        diferencia /= modulo;
        atom_pos.row(ind2) += (modulo - bond)*diferencia;

        // Adds the atom bonded to the extreme atom to the queue and remove them
        // from the to_move list.
        bonds_list = bonds_info[ind2];
        for (int i=0; i < bonds_list.size(); ++i){
            auto new_bond = bonds_list[i].first;
            if (to_move.find(new_bond) != to_move.end()){
                waiting.push(std::make_pair(ind2, bonds_list[i]));
                to_move.erase(new_bond);
            }
        }
    }
}


arma::rowvec find_atom_random_displ(const arma::mat &atom_pos,
                                    bonds &bonds_info, int atom_index,
                                    double sigma_scale){
    int Nbonded_ref = int(bonds_info[atom_index].size());

    double sigma = bonds_info[atom_index][0].second * sigma_scale;

    arma::rowvec direction(3);
    arma::rowvec dir_vec1(3);
    arma::rowvec dir_vec2(3);
    switch (Nbonded_ref) {
    case 1:
        dir_vec1 = arma::randu<arma::rowvec>(3);
        dir_vec2 = atom_pos.row(bonds_info[atom_index][0].first);
        dir_vec2 -= atom_pos.row(atom_index);
        break;
    case 2:
        dir_vec1 = arma::randu<arma::rowvec>(3);
        dir_vec2 = atom_pos.row(bonds_info[atom_index][0].first);
        dir_vec2 -= atom_pos.row(bonds_info[atom_index][1].first);
        break;
    default:
        dir_vec1 = atom_pos.row(bonds_info[atom_index][0].first);
        dir_vec1 -= atom_pos.row(bonds_info[atom_index][2].first);
        dir_vec2 = atom_pos.row(bonds_info[atom_index][0].first);
        dir_vec2 -= atom_pos.row(bonds_info[atom_index][1].first);
        break;
    }
    direction = arma::cross(dir_vec1, dir_vec2);
    direction /= arma::norm(direction);
    direction *= (random_direction(random_engine2)*2-1);
    direction *= sigma*random_normal(random_engine2);
    return direction;
}
